src/app/OrderBook.tsx

```diff
@@ -1,30 +1,219 @@
-import { useContext } from "react";
+import { useContext, CSSProperties } from "react";
 import { OrderbookLine } from "alphadex-sdk-js";
 import { AdexStateContext } from "./contexts";
+import "./orderbook.css";
+import * as utils from "./utils";
 
-export function OrderBook({ orders }: { orders: OrderbookLine[] }) {
-  //returns simple orderbook of buys/sells
+// TODO: test the table updates automatically when orders get bought
+
+export interface OrderBookRowProps {
+  barColor?: string;
+  orderCount?: number;
+  price?: number;
+  size?: number;
+  total?: number;
+  maxTotal?: number;
+  absentOrders?: string;
+}
+
+function OrderBookRow(props: OrderBookRowProps) {
+  const adexState = useContext(AdexStateContext);
+
+  const { barColor, orderCount, price, size, total, maxTotal } = props;
+  if (
+    typeof barColor !== "undefined" &&
+    typeof orderCount !== "undefined" &&
+    typeof price !== "undefined" &&
+    typeof size !== "undefined" &&
+    typeof total !== "undefined" &&
+    typeof maxTotal !== "undefined"
+  ) {
+    const { maxDigitsToken1, maxDigitsToken2 } = adexState.currentPairInfo;
+    const priceString = utils.displayNumber(price, maxDigitsToken1);
+    const sizeString = utils.displayNumber(size, maxDigitsToken2);
+    const totalString = utils.displayNumber(total, maxDigitsToken1);
+    const barWidth = `${(total / maxTotal) * 100}%`;
```

Comment:
@fliebenberg @SmashingBumpkin @steferr is that the correct way to calculate the bar width of the orderbook rows?

Reply:
Nice one! Looks good. The calculation and display both seem correct.

Reply:
Couple of comments:
1. Price is always displayed in terms of token2, so the correct maxDigits for price will be maxDigitsToken2. The way to read the price is: number of token2 I can get / have to pay for one token1. So a price of 100 means I have to pay 100token2 to get 1 token1.
2. Size is usually specified in token1. I.e. it tells you how many token1 is available at that price.
3. The total column is usually just the cumulative sum of the Size columns, so should also be specified in token1. If you want to show the number of token2 in a orderbookline, you can call the column "Value" and then specify the amount of token2 (so it will essentially be the equivalent of the "Size" column, but specified in token2. You could even have a "Total Value" column as well which would just be the "Total" column but specified in token2. This is not typically specified though.
4. Happy with the barWidth calc. Just a question: Is maxTotal the total of displayed lines or all lines (including the ones not displayed)? I think either can work but if I had to choose I would go with displayed lines.

Reply:
This is looking great already.
I think we need to start looking at merging some of this work into main, even if it is not completely done. We can then rather do individual issues to improve it. It is just easier when main is not too far out of sync with any particular branch, unless that branch is some new version of the app that we are purposefully keeping separate.

Reply:
Ok, fixing the columns, I did have my doubts about which should be which, thank you. `maxTotal` currently only takes into consideration the visible rows. Also there is no scrolling or aggregation, or any other means of getting the rows beyond the first n (currently 8) - which if I recall correctly is what we planned to have in the MVP.